src/java/util/Gradient.java

```diff
@@ -0,0 +1,52 @@
+package util;
+
+import lombok.AllArgsConstructor;
+
+import java.awt.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+public strictfp class Gradient {
+    public Map<Float, Color> colors = new TreeMap<>();
+
+    public void addColor(float period, Color color){
+        colors.put(period, color);
+    }
+
+    public Color evaluate(float period){
+        if (period > 1f || period < 0f){
+            throw new RuntimeException("Period must be comprised between 0 and 1 (supplied: "+period+")");
+        }
+
+        float previousKey = 0f;
+        float nextKey = 1f;
+        for (float key : colors.keySet()){
```

Comment:
you're using a red black tree for storing the map https://en.wikipedia.org/wiki/Red%E2%80%93black_tree . You can use this data structure to search for an index minimally greater or lower than the supplied one, see: https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#higherEntry-K-